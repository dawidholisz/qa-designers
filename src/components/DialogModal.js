import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { Formik } from 'formik'
import * as Yup from 'yup'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

const validationSchema = Yup.object().shape({
  location: Yup.string().required('Required'),
})

const DialogModal = ({ isVisible, setLocation, closeModal, header, description }) => {
  return (
    <div className={classNames('modal', { '-visible': isVisible })} data-testId="location-modal">
      <Formik
        initialValues={{ location: localStorage.getItem('weatherLocation') }}
        onSubmit={(values, { setSubmitting }) => {
          setTimeout(() => {
            setLocation(values.location)
            setSubmitting(false)
          }, 400)
        }}
        validationSchema={validationSchema}
      >
        {({ values, errors, touched, handleChange, handleBlur, handleSubmit, isSubmitting }) => (
          <div className="modal__dialog">
            <button className="close-button" data-testid="close-button" onClick={closeModal}>
              <FontAwesomeIcon icon="times" size="lg" />
            </button>

            <h1>{header}</h1>
            <form onSubmit={handleSubmit}>
              <div className="modal__description">
                <p>{description}</p>
              </div>
              <div className="modal__input">
                <label htmlFor="location">Localization:</label>
                <input
                  name="location"
                  onBlur={handleBlur}
                  onChange={handleChange}
                  placeholder="Weather Location"
                  type="text"
                  value={values.location}
                  data-testId="city-input"
                />
                {errors.location && touched.location && errors.location}
              </div>
              <button className="set-button" disabled={isSubmitting} type="submit" data-testid="save-city-btn">
                Set
              </button>
            </form>
          </div>
        )}
      </Formik>
    </div>
  )
}

DialogModal.propTypes = {
  isVisible: PropTypes.bool.isRequired,
  header: PropTypes.string.isRequired,
  description: PropTypes.string,
  setLocation: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
}

export default DialogModal
