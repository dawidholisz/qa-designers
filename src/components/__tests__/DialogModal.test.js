import React from 'react'
import { render, fireEvent, wait } from '@testing-library/react'
import DialogModal from '../DialogModal'

const renderComponent = props => {
  return render(<DialogModal {...props} />)
}
const props = {
  closeModal: jest.fn(x => x),
  header: 'header',
  isVisible: true,
  setLocation: jest.fn(x => x),
}

describe('<DialogModal/>', () => {
  it('render DialogModal with header and decription', () => {
    const { container } = renderComponent(props)
    expect(container.firstChild).toMatchSnapshot()
  })
  it('render DialogModal without -visible classname', () => {
    const { container } = renderComponent({ ...props, isVisible: false })
    expect(container.firstChild).toMatchSnapshot()
  })
  it('click close button on modal', () => {
    const { getByTestId } = renderComponent(props)
    fireEvent.click(getByTestId('close-button'))

    expect(props.closeModal).toHaveBeenCalled()
  })
  it('click set button on modal', async () => {
    const testLocation = 'TestLocation'
    const { getByText, getByPlaceholderText } = renderComponent(props)
    fireEvent.change(getByPlaceholderText('Weather Location'), {
      target: { value: testLocation },
    })
    fireEvent.click(getByText('Set'))
    await wait(() => {
      expect(props.setLocation).toHaveBeenCalled()
      expect(props.setLocation).toHaveBeenCalledWith(testLocation)
    })
  })
})
