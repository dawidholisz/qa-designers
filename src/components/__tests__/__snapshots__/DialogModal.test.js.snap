// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DialogModal/> render DialogModal with header and decription 1`] = `
<div
  class="modal -visible"
>
  <div
    class="modal__dialog"
  >
    <button
      class="close-button"
      data-testid="close-button"
    />
    <h1>
      header
    </h1>
    <form>
      <div
        class="modal__description"
      >
        <p />
      </div>
      <div
        class="modal__input"
      >
        <label
          for="location"
        >
          Localization:
        </label>
        <input
          name="location"
          placeholder="Weather Location"
          type="text"
          value=""
        />
      </div>
      <button
        class="set-button"
        type="submit"
      >
        Set
      </button>
    </form>
  </div>
</div>
`;

exports[`<DialogModal/> render DialogModal without -visible classname 1`] = `
<div
  class="modal"
>
  <div
    class="modal__dialog"
  >
    <button
      class="close-button"
      data-testid="close-button"
    />
    <h1>
      header
    </h1>
    <form>
      <div
        class="modal__description"
      >
        <p />
      </div>
      <div
        class="modal__input"
      >
        <label
          for="location"
        >
          Localization:
        </label>
        <input
          name="location"
          placeholder="Weather Location"
          type="text"
          value=""
        />
      </div>
      <button
        class="set-button"
        type="submit"
      >
        Set
      </button>
    </form>
  </div>
</div>
`;
